### DISCUSSION TITLE
Change Streams not working

### DISCUSSION THREAD

**Discussion by philkunz** [typegoose/mongodb-memory-server#642] [Q&A] [answered]
Any reason why the replica set created with mongodb-memory-server does not work for change streams in general?
while Atlas just works?

```
const changeStream = mongoDbCollection.watch([], {
    fullDocument: 'updateLookup'
});
```

**Comment by nodkz** [comment, marked as answer]
Please read requirements for streams in mongodb and use proper engine and other settings when you spin up your replica set
https://www.mongodb.com/docs/manual/changeStreams/#availability

If you get working solution, please share it here.
Tnx.

**Reply by hasezoey** [reply]
as a TL;DR: to the link provided:
- your storage engine must be `wiredTiger` (by default in MMS it is `ephermeralForTest`)
- if you use mongodb before 4.2 (not including 4.2), you will have to set the read-concern to `majority`

**Reply by philkunz** [reply]
I confirm. Simply switching to wiredTiger as storage engine works.

**Reply by sibelius** [reply]
do you have a test example of change stream working ?

**Reply by noghartt** [reply]
See below an example to replace the `storageEngine` for `wiredTiger`:

```ts
  await MongoMemoryReplSet.create({
    instanceOpts: [{ storageEngine: 'wiredTiger' }, { storageEngine: 'wiredTiger' }],
    binary: {
      systemBinary: process.env.MONGODB_BINARY,
      skipMD5: true,
    },
    replSet: {
      count: 2,
    },
  });
```

You can put the `storageEngine` inside `replSet` too:

```ts
  await MongoMemoryReplSet.create({
    binary: {
      systemBinary: process.env.MONGODB_BINARY,
      skipMD5: true,
    },
    replSet: {
      count: 2,
      storageEngine: 'wiredTiger',
    },
  });
```

**Reply by diegosasw** [reply]
So, is it possible to have a single docker image with mongodb-memory-server spinning up a replica set so that the change streams work?

I created a docker image (See this https://gitlab.com/sunnyatticsoftware/sandbox/mongo-rs and https://stackoverflow.com/q/76335421/2948212)  that contains a replica set (3 nodes) but I'd like it to be in-memory persistence, and I found this project.

I didn't understand if this is currently supported?

**Reply by hasezoey** [reply]
> So, is it possible to have a single docker image with mongodb-memory-server spinning up a replica set so that the change streams work?

yes, you just need to explicitly set the storage engine used, because the default (in mongodb versions before 7.0) is set to `ephermeralForTest` in mongodb-memory-server, which does not support change streams
